src/charm.py

```diff
@@ -1,217 +1,416 @@
+            config_hash=None,
+            launched_with_peers=False,
+        )
 
-    def build_config_file(self):
-        """Create the alertmanager config file from self.model.config"""
-        if not self.model.config["pagerduty_key"]:
-            raise BlockedStatusError("Missing pagerduty_key config value")
+    @property
+    def num_peers(self) -> int:
+        """Number of peer units (excluding self)
+        """
+        # For some reason in Juju 2.9.5 `self.peer_relation.units` is an empty set
+        return sum(
+            isinstance(unit, ops.model.Unit) and unit is not self.unit
+            for unit in self.peer_relation.data.keys()
+        )
 
-        return CONFIG_CONTENT.format(pagerduty_key=self.model.config["pagerduty_key"])
+    @property
+    def peer_relation(self) -> ops.model.Relation:
+        return self.model.get_relation(self._peer_relation_name)
+
+    @property
+    def private_address(self) -> Optional[str]:
+        """Get the unit's ip address.
+        Technically, receiving a "joined" event guarantees an IP address is available. If this is
+        called beforehand, a None would be returned.
+        When operating a single unit, no "joined" events are visible so obtaining an address is a
+        matter of timing in that case.
+
+        This function is still needed in Juju 2.9.5 because the "private-address" field in the
+        data bag is being populated by the app IP instead of the unit IP.
+        Also in Juju 2.9.5, ip address may be None even after RelationJoinedEvent, for which
+        "ops.model.RelationDataError: relation data values must be strings" would be emitted.
+
+        Returns:
+          None if no IP is available (called before unit "joined"); unit's ip address otherwise
+        """
+        # if bind_address := check_output(["unit-get", "private-address"]).decode().strip()
+        if bind_address := self.model.get_binding(self.peer_relation).network.bind_address:
+            bind_address = str(bind_address)
+        return bind_address
+
+    def _store_private_address(self):
+        """Store private address in unit's peer relation data bucket.
+
+        This function is still needed in Juju 2.9.5 because the "private-address" field in the
+        data bag is being populated by the app IP instead of the unit IP.
+        Also in Juju 2.9.5, ip address may be None even after RelationJoinedEvent, for which
+        "ops.model.RelationDataError: relation data values must be strings" would be emitted.
+        """
+        self.peer_relation.data[self.unit]["private_address"] = self.private_address
+
+    def _fetch_private_address(self) -> Optional[str]:
+        """Fetch private address from unit's peer relation data bucket.
+        """
+        return self.peer_relation.data[self.unit].get("private_address")
+
+    def _alertmanager_layer(self) -> Dict[str, Any]:
+        """Returns Pebble configuration layer for alertmanager.
+        """
+
+        def _command():
```

Comment:
Why is this defined inside `_alertmanager_layer` rather than being its own method?

Reply:
For encapsulation/hiding reasons: it is only needed there and should probably be never used in other context.

Reply:
This feels very counter to standard practice. _command is not really hidden from anything inside there and its weird to have the method redefined on every call to _alertmanager_layer (and maybe inefficient but I don't know)

Reply:
Nested functions are very common in factories and closures.
Encapsulation is also common.
Another way to encapsulate the same would be e.g. `class AlertmanagerPebbleLayer`.

Reply:
okay